Block Bindings: Change placeholder when attribute is bound

SantosGuillamot opened the pull request:
## What?
Adapt the placeholder and the `aria-label` attribute of the rich text component when the relevant attribute is connected to a block bindings source. The default texts don't make sense in this context.

| Before | After |
|----------|----------|
| ![Screenshot 2024-08-29 at 14 34 50](https://github.com/user-attachments/assets/747b91fd-924b-453a-ba57-bb767745d21e) | ![Screenshot 2024-08-29 at 14 33 44](https://github.com/user-attachments/assets/751f6ade-ca50-49a8-9036-32f7b183bfb5) |

Same happens with the `aria-label` text, which is also changed in this PR.

Additionally, I took the opportunity to do a small refactor on `disableBoundBlock`, which is used to lock editing when appropriate, to rely on the rich-text identifier instead of iterating through all the attributes.

## Why?
It is confusing to get a prompt to insert a block when user is actually editing the binding source value.

## How?
Return an alternative placeholder when the attribute used as identifier in the RichText is connected to any source. And use this placeholder when it exists.

## Testing Instructions
1. Register a new custom field with an empty string as default value.
```
	register_meta(
		'post',
		'empty_field',
		array(
			'show_in_rest' => true,
			'single'       => true,
			'type'         => 'string',
			'default'      => '',
		)
	);
```
2. Go to a page and insert a paragraph bound to that custom field.
```
<!-- wp:paragraph {"metadata":{"bindings":{"content":{"source":"core/post-meta","args":{"key":"empty_field"}}}}} -->
<p></p>
<!-- /wp:paragraph -->
```
3. Check that the placeholder shows the new text related to bindings.
4. Inspect the element and check the `aria-label` has also changed.
5. Add an empty paragraph (not bound) and check it keeps working as expected.

SantosGuillamot committed:
Change placeholder when attribute is bound

SantosGuillamot commented on packages/block-editor/src/components/rich-text/index.js:
```diff
@@ -163,49 +163,41 @@ export function RichTextWrapper(
+					args: relevantBinding.args,
+				} );
+
+			const _bindingsPlaceholder =
+				( relevantBinding?.args?.key || blockBindingsSource?.label ) +
+				' value is empty';
```
It'd be great to get feedback on the desired text.

cbravobernal replied:
I think is fine, maybe we can remove `value`.

SantosGuillamot replied:
I thought about removing it, but when using the source label felt a bit weird:

* "my_custom_field is empty" sounds fine.
* "Custom Source is empty" doesn't sound as good to me.

Having said that, I don't have a strong opinion on this.

cbravobernal replied:
Me neither 🤷‍♂️

SantosGuillamot committed:
Only change placeholder when value is empty

cbravobernal commented on packages/block-editor/src/components/rich-text/index.js:
```diff
@@ -163,49 +163,43 @@ export function RichTextWrapper(
+					args: relevantBinding.args,
+				} );
+
+			const _bindingsPlaceholder =
+				( relevantBinding?.args?.key || blockBindingsSource?.label ) +
+				' value is empty';
```
```suggestion
			const _bindingsPlaceholder =
			// Translators: %s is the bindings source or bindings key placeholder.
			 sprintf( '%s value is empty', ( relevantBinding?.args?.key || blockBindingsSource?.label )  );
```

cbravobernal replied:
This should be translatable.

SantosGuillamot replied:
You're totally right. I just commited your suggestion.

SantosGuillamot committed:
Translate empty text

Co-authored-by: Carlos Bravo <37012961+cbravobernal@users.noreply.github.com>

SantosGuillamot committed:
Adapt latest commit

SantosGuillamot committed:
Add `adjustedValue` in `useSelect`

gziolo commented on packages/block-editor/src/components/rich-text/index.js:
```diff
@@ -163,49 +164,45 @@ export function RichTextWrapper(
+
+			const _bindingsPlaceholder = sprintf(
+				/* translators: %s: source label or key */
+				__( '%s value is empty' ),
```
It should be a prompt for action, example:

```suggestion
				__( 'Type something to provide value for %s' ),
```

It should be rather more concise.

SantosGuillamot replied:
There are two aspects to consider here:

* To simplify the workflows and the implementation, my idea was to reuse the same "empty" message when the user can edit and can not edit the meta field. For example, we are showing this message when you bind a paragraph to an empty field in a query loop, where editing is disabled.
* This message appears in every rich text, including shorter blocks like "Button". With that in mind, I would like to keep it as short as possible.

Do you think that showing " X value is empty" is not enough for prompting the user to write? Do you think that we should show different messages when the user can edit or can not?

gziolo replied:
Yes, these are two different use cases. In particular, these placeholder prompts calling for action were designed with accessibility in mind. It's best to check it with assistive technology to check whether it makes sense.

SantosGuillamot replied:
Okay, I have just pushed a commit to differentiate between these use cases. It is now a matter of deciding the text we want to show for each case. What about this:

Use "Add {key} value" when the user can edit:

![Screenshot 2024-08-30 at 15 05 52](https://github.com/user-attachments/assets/09184a27-d8d1-4a88-ac80-a4892858bdfe)

This follows what we are doing for post title or buttons for example. It is simple, but not sure if it is clear enough.

![Screenshot 2024-08-30 at 15 05 02](https://github.com/user-attachments/assets/32be86e7-9c5d-448b-8010-56018ef8d406)

Use just the key (or label) when the editing is locked.

![Screenshot 2024-08-30 at 15 06 15](https://github.com/user-attachments/assets/65da2c90-0cb5-4518-af84-bcd99dc700c4)

cbravobernal replied:
Maybe just "Add empty_field"? Similar to add title, I think value is still not adding "value" 😆

SantosGuillamot replied:
That could work for me 😄

SantosGuillamot committed:
Different message depending on capabilities

cbravobernal commented on packages/block-editor/src/components/rich-text/index.js:
```diff
@@ -404,9 +403,11 @@ export function RichTextWrapper(
 				aria-readonly={ shouldDisableEditing }
 				{ ...props }
+				aria-label={
+					bindingsPlaceholder || props[ 'aria-label' ] || placeholder
```
is `placeholder` already defaulted on line 337?

artemiomorales replied:
> is placeholder already defaulted on line 337?

+1 — I think having just `aria-label={ placeholder || props[ 'aria-label' ] }` should be good.

SantosGuillamot replied:
The problem is the priority. `props[ 'aria-label' ]` is supposed to override the default placeholder value if it exist for the `aria-label` attribute. And I believe we want to override the `aria-label` prop when a binding exists.

gziolo replied:
@zaguiini, regarding your comment from https://github.com/WordPress/gutenberg/pull/65089#pullrequestreview-2283038208.

> Another issue, though related, is that the Block Bindings API does not respect an existing placeholder attribute. From the back-end, I'm setting a bespoke placeholder:

I think this is what changed the behavior, as placeholder became here a fallback. Now, the question is whether there is always `placeholder` coming from the server. If not, then we should be fine reverting to the old behavior and including the fallback to `bindingsPlaceholder` when the RichText is connected to a source.

cbravobernal commented on packages/block-editor/src/components/rich-text/index.js:
```diff
@@ -163,49 +164,47 @@ export function RichTextWrapper(
 			}
 
-			return _disableBoundBlocks;
+			const relevantBinding = blockBindings[ identifier ];
```
I don't really like the name, but cannot find any better. Why is relevant?

artemiomorales replied:
I find `relevantBinding` to be unclear.

Maybe `relatedBinding` could work? Another option is `associatedBinding`.
If neither of those, honestly for me just `binding` would work fine and might even be the clearest. Another possibility is `richTextBinding` (my least favorite, but could work). 

Other ideas: `connectedBinding`, `targetBinding`, `linkedBinding`, though I like these options less and they also seem unclear to me.

SantosGuillamot replied:
I can use `relatedBinding` 👍

artemiomorales reviewed (commented):
Looking good! I tested and the placeholder shows as expected in the post editor and site editor. As far as I can tell, this should be good to go after addressing the points below.

SantosGuillamot committed:
Remove "empty" from placeholder

SantosGuillamot committed:
Change `relatedBinding` variable name

SantosGuillamot commented:
I believe I've addressed the feedback you shared. This is how it is looking at this moment:

**Empty value when user CAN edit**

![Screenshot 2024-09-02 at 12 21 20](https://github.com/user-attachments/assets/d7ad3030-6dcd-4bec-89bd-5fae35716132)


**Empty value when user CAN NOT edit**

![Screenshot 2024-09-02 at 12 21 59](https://github.com/user-attachments/assets/7dac1484-d7d3-4368-b502-cb8aa4f4c867)

Is this good enough for the first version? Something else I should change?

cbravobernal reviewed (approved):
LGTM
We can merge it and wait for more feedback from the community.
